lib/drizzle/io.ex

```diff
@@ -44,22 +44,29 @@ defmodule Drizzle.IO do
   defp init_output(pin) do
     {:ok, gpio} = Circuits.GPIO.open(pin, :output)
     :ok = Circuits.GPIO.write(gpio, 1)
-    %{gpio: gpio, currstate: 0}
+    %{gpio: gpio, currstate: false}
   end
 
-  defp do_status_change(zone_name, zonestruct, newstate) do
-    intstate = (newstate && 1 || 0)
-    DrizzleUiWeb.Endpoint.broadcast @topic, "zone status change", %{zone: zone_name, newstate: intstate}
-    :ok = Circuits.GPIO.write(zonestruct.gpio, (!newstate && 1 || 0))
-    intstate
+  # map falsey values to a 0 or 1, as needed by Circuits.GPIO NIF
+  defp intstate(state) when state in [false, nil], do: 0
+  defp intstate(_state), do: 1
+
+  defp do_status_change(zone_name, zonestruct, desiredstate) do
+    DrizzleUiWeb.Endpoint.broadcast @topic, "zone status change", %{zone: zone_name, newstate: desiredstate}
+    # GPIO.write(0) actually turns ON the relay on the Waveshare RPi 8x relay board
+    # this has to do with pull-up or down resitors, we might need to make this configurable
+    :ok = Circuits.GPIO.write(zonestruct.gpio, intstate(!desiredstate))
+    desiredstate
   end
 
   def handle_cast({:activate, zone}, state) do
     {:noreply,
-    # activate this zone, but turn off all other zones
-     state
-     |> Enum.map(fn {zone_name, %{gpio: gpio, currstate: _cst} = zonestruct} ->
-       {zone_name, %{gpio: gpio, currstate: do_status_change(zone_name, zonestruct, zone_name == zone) }}
+    state
+    |> Enum.map(fn {zone_name, %{gpio: gpio, currstate: _cst} = zonestruct} -> {
```

Comment:
One last thing (I learned this recently), it's not preferable to do pipes if you only have one of them, just turn this into a normal call to Enum.map(state, ...). Also, why do you pattern match on the map field currstate if you are not going to use it (and you don't have a second case)? You could simplify this as:

```elixir
@doc """
Activate this zone, turn off all other zones
"""
def handle_cast({:activate, zone}, state) do
  {:noreply, toggle_zones(state, zone)}
end

defp toggle_zones(state, zone) do
  Enum.map(state, & toggle_zone(&1, zone))
end

defp toggle_zone({zone_name, zonestruct}, zone) do
  {
    zone_name,
    %{
       gpio: zonestruct.gpio,
       currstate: do_status_change(zone_name, zonestruct, zone_name == zone)
    }
  }
end
```

This is all optional, I just like to break things into small functions (I might overdo it sometimes, so take this with a grain of salt).

Reply:
No, you can't use 'toggle' in this way as I only want to deactive other zones when activating a specific zone. I don't want to adtivate all zones when I deactivate a specific zone.

Reply:
I will fix the pipe issue now

Reply:
My intention was for the code to be equivalent to what you had, did I make a mistake?

Reply:
activate(2) -> activate zone 2, deactivate 1,3,4,5,6,7,8
deactivate(2) -> only deactivate 2